clash-prelude/src/Clash/Class/BitPack/Internal.hs

```diff
@@ -1,102 +1,441 @@
 {-|
-Copyright  :  (C) 2019, QBayLogic B.V.
+Copyright  :  (C) 2013-2016, University of Twente,
+                  2016-2017, Myrtle Software Ltd
 License    :  BSD2 (see the file LICENSE)
 Maintainer :  Christiaan Baaij <christiaan.baaij@gmail.com>
 -}
+
 {-# LANGUAGE CPP #-}
+{-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# LANGUAGE Trustworthy #-}
+
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.Extra.Solver #-}
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.Normalise #-}
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.KnownNat.Solver #-}
+{-# OPTIONS_HADDOCK show-extensions #-}
+
+#include "MachDeps.h"
+
+module Clash.Class.BitPack.Internal where
+
+import Control.Exception              (catch, evaluate)
+import Data.Binary.IEEE754            (doubleToWord, floatToWord, wordToDouble,
+                                       wordToFloat)
+
+import Data.Complex                   (Complex)
+import Data.Int
+import Data.Ord                       (Down)
+import Data.Word
+import Foreign.C.Types                (CUShort)
+import GHC.TypeLits                   (KnownNat, Nat, type (+), type (-))
+import Numeric.Half                   (Half (..))
+import GHC.Generics
+import GHC.TypeLits.Extra             (CLog, Max)
+import Prelude                        hiding (map)
+import System.IO.Unsafe               (unsafeDupablePerformIO)
+
+import Clash.Promoted.Nat             (SNat(..), snatToNum)
+import Clash.Annotations.Primitive    (hasBlackBox)
+import Clash.Class.BitPack.Internal.TH (deriveBitPackTuples)
+import Clash.Class.Resize             (zeroExtend, resize)
+import Clash.Sized.Internal.BitVector
+  (BitVector, Bit, pack#, split#, checkUnpackUndef, undefined#, unpack#,
+   unsafeToNatural, (++#), isLike#)
+import Clash.XException
+
+{- $setup
+>>> :set -XDataKinds
+>>> import Clash.Prelude
+-}
+
+-- | Convert to and from a 'BitVector'
```

Comment:
I wonder if it would be worth expanding on the docs? Give some rationale for users about what types can be converted to / from BV (i.e. why is there no list instance), or why you might even want to do it at all from a Haskell POV?